### ISSUE TITLE
An input channel evaluating to null results in a misleading error message

### ISSUE THREAD

**Issue by ewels** [nextflow-io/nextflow#242]
### Working test case

`minimal.nf`:
```groovy
params.download_fasta = false
process downloadFASTA {
    input: val dl_url from params.download_fasta
    when: params.download_fasta
    script:
    """
    curl -O -L $dl_url
    """
}
```

`nextflow.config`:
```groovy
params.bar = 'foo'
process {
  foo = params.bar
}
```
Works as expected:
```
$ nextflow run minimal.nf
N E X T F L O W  ~  version 0.22.4
Launching `minimal.nf` [agitated_hodgkin] - revision: b3b3101f7c
[warm up] executor > local
```

### Breaking test case
Make this change to`nextflow.config`:
```groovy
// params.bar = 'foo'
process {
  foo = params.bar
}
```
Now breaks:
```
$ nextflow run minimal.nf
N E X T F L O W  ~  version 0.22.4
Launching `minimal.nf` [tiny_nobel] - revision: b3b3101f7c
[warm up] executor > local
ERROR ~ No such variable: dl_url

 -- Check script 'minimal.nf' at line: 2 or see '.nextflow.log' file for more details
```

### Working again:
Weirdly, changing the syntax as follows fixes it again:
`nextflow.config`:
```groovy
// params.bar = 'foo'
process.foo = params.bar
```

```
$ nextflow run minimal.nf
N E X T F L O W  ~  version 0.22.4
Launching `minimal.nf` [compassionate_kalam] - revision: b3b3101f7c
[warm up] executor > local
```

Any ideas what's going on here? @viklund also testing, will probably follow with his own minimal example..

**Comment by viklund** [comment]
These are my two files:

    $ cat nextflow.config
    //params.bar = ''
    process {
      foo = params.bar
    }
    
    $ cat test.nf
    //params.value = "hola"
    process buggy {
        input:
            val v from params.value

        script:
        """
        echo $v
        """
    }

The example requires running these without anything specified on the command line (that is, no `--value 5` or `--bar 3`).

when running this fails. But if both commented lines are inserted, it works (but not with any one of them). The error message is the same as @ewels got:

    $ nextflow run test.nf
    N E X T F L O W  ~  version 0.22.4
    Launching `test.nf` [pensive_nobel] - revision: 71e67de276
    [warm up] executor > local
    ERROR ~ No such variable: v
    
     -- Check script 'test.nf' at line: 2 or see '.nextflow.log' file for more details

**Comment by pditommaso** [comment]
OK, it can be reduced to this 

```
process buggy {
    input:
    val v from null

    script:
    """
    echo $v
    """
}
```

**Comment by pditommaso** [comment]
This brings to an interesting question? Should a not defined `params` return `null` or throw an exception?

**Comment by ewels** [comment]
But does your example also cover my failure above where I'm not changing anything to do with the process? There `params.download_fasta` is not `null`, it's `false`...

**Comment by ewels** [comment]
Same error message, but I suspect that the mode of failure is different (probably why we found testing this so confusing this morning). Could it be anything to do with #48?

**Comment by ewels** [comment]
And in answer to your question - I think throwing an exception is fine, as long as `when:` returns `false` when the input is `null` so that the process is skipped. That was the behaviour that I expected and was trying to achieve...

**Comment by pditommaso** [comment]
> But does your example also cover my failure above where I'm not changing anything to do with the process? There params.download_fasta is not null, it's false

This is slight different. The main problem is that the following snippet is assigning to `foo` a non existing parameter. For some causes that I'm still investigating that corrupts the process definition. 

I've added a more strict check warning on that, however the general advice is to define **always** the parameter before use them. 

I've added a new snapshot, you may want to check if solve your problem. You can try it by using: 

```
NXF_VER=0.22.5-SNAPSHOT nextflow run ... etc
```

**Comment by ewels** [comment]
ok, so the thing to remember is to define `param` variables in config files before using them.. Makes sense I think. The use case I had was as follows:

```groovy
process {
  clusterOptions: "-A $params.project"
}
```

So I should do the following:
```groovy
params {
  project = false
}
process {
  clusterOptions: "-A $params.project"
}
```
I'll try this now. But specifying `--project` on the command line should still work in this case? When is `clusterOptions` evaluated?

**Comment by pditommaso** [comment]
Yes, parameter priority is the following (from higher to lower)

* command line
* config file
* script file

**Comment by pditommaso** [comment]
Included in version `0.22.5`

**Comment by ewels** [comment]
To follow up, my example above was a little wrong as I was missing `{}` around the `clusterOptions`. Once I had those, it was evaluated later in the pipeline after `params.project` had been updated by other config files, and it worked.
